File: nwerc2024/evolvingetymology/submissions/accepted/Maarten.kt
fun main() {
    var (n, k) = readln().split(" ").map { it.toLong() }
    val s = readln()

    var offset: Long = 1
    var base: Long = 2
    while (k > 0) {
        if ((k and 1) == 1L) {
            offset = (offset * base) % n
        }
        k = k shr 1
        base = (base * base) % n
    }

    println((0 until n).map { s[((it * offset) % n).toInt()] }.joinToString(""))
}

File: nwerc2024/evolvingetymology/submissions/accepted/maarten.cpp
#include <iostream>
#include <string>
#include <vector>

using namespace std;

int main() {
    long long n, k;
    cin >> n >> k;
    string s;
    cin >> s;

    long long offset = 1, base = 2;
    while (k > 0) {
        if (k & 1) {
            offset = (offset * base) % n;
        }
        k >>= 1;
        base = (base * base) % n;
    }

    string result;
    for (int i = 0; i < n; ++i) {
        result += s[(i * offset) % n];
    }

    cout << result << endl;
    return 0;
}
